File: scala/protein-translation/src/main/scala/ProteinTranslation.scala
object ProteinTranslation {
  private val translations = Map("AUG" -> "Methionine", "UUU" -> "Phenylalanine", "UUC" -> "Phenylalanine", "UUA" -> "Leucine", "UUG" -> "Leucine", "UCU" -> "Serine", "UCC" -> "Serine", "UCA" -> "Serine", "UCG" -> "Serine", "UAU" -> "Tyrosine", "UAC" -> "Tyrosine", "UGU" -> "Cysteine", "UGC" -> "Cysteine","UGG" -> "Tryptophan", "UAA" -> "STOP", "UAG" -> "STOP", "UGA" -> "STOP")
  def proteins(strand: String): List[String] = {
    strand.grouped(3).map(translations(_)).span(_ != "STOP")._1.toList
  }
}

File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

chunked :: Int -> [a] -> [[a]]
chunked _ [] = []
chunked n l = take n l : chunked n (drop n l)

convert :: String -> Maybe String
convert x
  | x == "AUG" = Just "Methionine"
  | x `elem` ["UUU", "UUC"] = Just "Phenylalanine"
  | x `elem` ["UUA", "UUG"] = Just "Leucine"
  | x `elem` ["UCU", "UCC", "UCA", "UCG"] = Just "Serine"
  | x `elem` ["UAU", "UAC"] = Just "Tyrosine"
  | x `elem` ["UGU", "UGC"] = Just "Cysteine"
  | x == "UGG" = Just "Tryptophan"
  | x `elem` ["UAA", "UAG", "UGA"] = Just "STOP"
  | otherwise = Nothing

proteins :: String -> Maybe [String]
proteins = fmap (takeWhile (/= "STOP")) . traverse convert . chunked 3
